File: backend/api/scans.py
"""
Scans API Blueprint.

Routes for interface scans and optical power data.
"""

from flask import Blueprint, request, jsonify
from ..utils.responses import success_response, error_response, list_response
from ..utils.errors import AppError, NotFoundError, ValidationError


scans_bp = Blueprint('scans', __name__, url_prefix='/api/scans')


def get_scan_service():
    """Get scan service instance."""
    from database import DatabaseManager
    from ..repositories.scan_repo import ScanRepository, OpticalPowerRepository
    from ..repositories.device_repo import DeviceRepository
    from ..services.scan_service import ScanService
    
    db = DatabaseManager()
    scan_repo = ScanRepository(db)
    optical_repo = OpticalPowerRepository(db)
    device_repo = DeviceRepository(db)
    return ScanService(scan_repo, optical_repo, device_repo)


@scans_bp.errorhandler(AppError)
def handle_app_error(error):
    """Handle application errors."""
    return jsonify(error_response(error.code, error.message, error.details)), error.status_code


@scans_bp.errorhandler(Exception)
def handle_generic_error(error):
    """Handle unexpected errors."""
    return jsonify(error_response('INTERNAL_ERROR', str(error))), 500


@scans_bp.route('/interfaces/<ip_address>', methods=['GET'])
def get_device_interfaces(ip_address):
    """
    Get latest interface scan data for a device.
    
    Args:
        ip_address: Device IP address
    
    Returns:
        List of interface records
    """
    service = get_scan_service()
    interfaces = service.get_device_interfaces(ip_address)
    
    return jsonify(list_response(interfaces))


@scans_bp.route('/optical', methods=['GET'])
def get_optical_interfaces():
    """
    Get all optical interfaces.
    
    Query params:
        ip_address: Optional device filter
    
    Returns:
        List of optical interfaces
    """
    service = get_scan_service()
    
    ip_address = request.args.get('ip_address')
    interfaces = service.get_optical_interfaces(ip_address)
    
    return jsonify(list_response(interfaces))


@scans_bp.route('/optical/power-history', methods=['GET'])
def get_optical_power_history():
    """
    Get optical power history for a device/interface.
    
    Query params:
        ip_address: Device IP address (required)
        interface_index: Optional interface filter
        hours: Time window in hours (default: 24)
    
    Returns:
        List of power readings
    """
    service = get_scan_service()
    
    ip_address = request.args.get('ip_address')
    if not ip_address:
        raise ValidationError('ip_address is required', field='ip_address')
    
    interface_index = request.args.get('interface_index')
    if interface_index:
        interface_index = int(interface_index)
    
    hours = int(request.args.get('hours', '24'))
    
    history = service.get_optical_power_history(ip_address, interface_index, hours)
    
    return jsonify(list_response(history))


@scans_bp.route('/optical/power-history', methods=['POST'])
def get_optical_power_history_bulk():
    """
    Get optical power history for multiple devices.
    
    Body:
        ip_addresses: List of device IP addresses
        interface_index: Optional interface filter
        hours: Time window in hours (default: 24)
    
    Returns:
        Dictionary mapping IP to power readings
    """
    service = get_scan_service()
    data = request.get_json() or {}
    
    ip_addresses = data.get('ip_addresses', [])
    if not ip_addresses:
        raise ValidationError('ip_addresses is required', field='ip_addresses')
    
    interface_index = data.get('interface_index')
    hours = data.get('hours', 24)
    
    history = service.get_power_history_for_devices(ip_addresses, interface_index, hours)
    
    return jsonify(success_response(history))


@scans_bp.route('/optical/trends', methods=['GET'])
def get_optical_power_trends():
    """
    Get optical power trend statistics.
    
    Query params:
        ip_address: Device IP address (required)
        interface_index: Interface index (required)
        days: Time window in days (default: 7)
    
    Returns:
        Trend statistics
    """
    service = get_scan_service()
    
    ip_address = request.args.get('ip_address')
    interface_index = request.args.get('interface_index')
    
    if not ip_address:
        raise ValidationError('ip_address is required', field='ip_address')
    if not interface_index:
        raise ValidationError('interface_index is required', field='interface_index')
    
    days = int(request.args.get('days', '7'))
    
    trends = service.get_optical_power_trends(ip_address, int(interface_index), days)
    
    return jsonify(success_response(trends))


@scans_bp.route('/cleanup', methods=['POST'])
def cleanup_old_data():
    """
    Clean up old scan and power data.
    
    Body:
        optical_days: Age threshold for optical data (default: 90)
    
    Returns:
        Cleanup statistics
    """
    service = get_scan_service()
    data = request.get_json() or {}
    
    optical_days = data.get('optical_days', 90)
    
    result = service.cleanup_old_data(optical_days)
    
    return jsonify(success_response(result, message='Cleanup completed'))

File: frontend/src/api/scans.js
/**
 * Scans API module.
 * 
 * Provides functions for interface scans and optical power data.
 */

import { apiClient, extractData, extractList } from './client';

const BASE_PATH = '/scans';

/**
 * Get interfaces for a device
 * @param {string} ipAddress - Device IP address
 * @returns {Promise<Array>} List of interfaces
 */
export async function getDeviceInterfaces(ipAddress) {
  const response = await apiClient.get(`${BASE_PATH}/interfaces/${ipAddress}`);
  return extractList(response);
}

/**
 * Get all optical interfaces
 * @param {string} ipAddress - Optional device filter
 * @returns {Promise<Array>} List of optical interfaces
 */
export async function getOpticalInterfaces(ipAddress = null) {
  const params = {};
  if (ipAddress) params.ip_address = ipAddress;
  
  const response = await apiClient.get(`${BASE_PATH}/optical`, params);
  return extractList(response);
}

/**
 * Get optical power history
 * @param {string} ipAddress - Device IP address
 * @param {Object} options - Options
 * @param {number} options.interfaceIndex - Interface index
 * @param {number} options.hours - Time window in hours
 * @returns {Promise<Array>} List of power readings
 */
export async function getOpticalPowerHistory(ipAddress, options = {}) {
  const params = { ip_address: ipAddress };
  if (options.interfaceIndex !== undefined) params.interface_index = options.interfaceIndex;
  if (options.hours) params.hours = options.hours;
  
  const response = await apiClient.get(`${BASE_PATH}/optical/power-history`, params);
  return extractList(response);
}

/**
 * Get optical power history for multiple devices
 * @param {Array<string>} ipAddresses - List of device IP addresses
 * @param {Object} options - Options
 * @param {number} options.interfaceIndex - Interface index
 * @param {number} options.hours - Time window in hours
 * @returns {Promise<Object>} Dictionary mapping IP to power readings
 */
export async function getOpticalPowerHistoryBulk(ipAddresses, options = {}) {
  const body = { ip_addresses: ipAddresses };
  if (options.interfaceIndex !== undefined) body.interface_index = options.interfaceIndex;
  if (options.hours) body.hours = options.hours;
  
  const response = await apiClient.post(`${BASE_PATH}/optical/power-history`, body);
  return extractData(response);
}

/**
 * Get optical power trends
 * @param {string} ipAddress - Device IP address
 * @param {number} interfaceIndex - Interface index
 * @param {number} days - Time window in days
 * @returns {Promise<Object>} Trend statistics
 */
export async function getOpticalPowerTrends(ipAddress, interfaceIndex, days = 7) {
  const params = {
    ip_address: ipAddress,
    interface_index: interfaceIndex,
    days,
  };
  
  const response = await apiClient.get(`${BASE_PATH}/optical/trends`, params);
  return extractData(response);
}

/**
 * Clean up old scan data
 * @param {number} opticalDays - Age threshold for optical data
 * @returns {Promise<Object>} Cleanup statistics
 */
export async function cleanupOldData(opticalDays = 90) {
  const response = await apiClient.post(`${BASE_PATH}/cleanup`, { optical_days: opticalDays });
  return extractData(response);
}

/**
 * Get scan progress
 * @returns {Promise<Object>} Scan progress status
 */
export async function getProgress() {
  const response = await apiClient.get('/progress');
  return response;
}

/**
 * Start a network scan
 * @returns {Promise<Object>} Scan start result
 */
export async function startScan() {
  const response = await apiClient.post('/scan');
  return response;
}

/**
 * Cancel running scan
 * @returns {Promise<Object>} Cancel result
 */
export async function cancelScan() {
  const response = await apiClient.post('/cancel_scan');
  return response;
}

export default {
  getDeviceInterfaces,
  getOpticalInterfaces,
  getOpticalPowerHistory,
  getOpticalPowerHistoryBulk,
  getOpticalPowerTrends,
  cleanupOldData,
  getProgress,
  startScan,
  cancelScan,
};
